### ISSUE TITLE
Need To Await Stdout

### ISSUE THREAD

**Issue by mpfaff** [dart-lang/sdk#39594]
I'm trying to present prompts in the console asynchronously on the Dart VM so that other tasks can run in the background. It seems that `stdout.write` is blocked by `await stdinLined.first`.

```dart
final Stream<String> stdinUtf8 = sharedStdIn.transform(utf8.decoder).asBroadcastStream();
final Stream<String> stdinLined = stdinUtf8.transform(LineSplitter());
```

```dart
Future<T> prompt<T>(String question, {T fallback, T def, String hint, T Function(String) parse}) async {
  assert(T is String || parse != null, '`parse` can only be null if T is String. T is $T');
  stdout.write('$question${hint != null || def != null ? ' (${hint ?? def})' : ''} >>> ');
  // await stdout.flush();
  final ret = await stdinLined.first;
  print('[OUTPUT = $ret]'); // for debugging
  if (ret.isEmpty && fallback != null) return def;
  else if (parse != null) {
    try {
      return parse(ret);
    } on FormatException {
      return fallback;
    }
  }

  return ret as T;
}
```

I've also noticed that `await stdout.flush()` does not actually wait for the contents of stdout to be printed to the console. Is there something else I should be doing?

**Example:**
```dart
Future<void> main() async {
  final name = await prompt<String>('What is your name?');
}
```

**Expected Output:**
```sh
$ dart bin/main.dart
What is your name? >>> <some user input>
[OUTPUT = <some user input>]
```

**Actual Output:**
```sh
$ dart bin/main.dart
<some user input>
What is your name? >>> [OUTPUT = <some user input>]
```

**Comment by mpfaff** [comment]
Bumping this issue

**Comment by sortie** [comment]
Hi @mpfaff 

It would've made things easier if you had provided a full standalone minimal example with all the imports and such, but I managed to piece your fragments together so no worries.

```dart
import 'dart:convert';
import 'dart:io';

import 'package:io/io.dart';

final Stream<String> stdinUtf8 = stdin.transform(utf8.decoder).asBroadcastStream();
final Stream<String> stdinLined = stdinUtf8.transform(LineSplitter());

Future<T> prompt<T>(String question, {T fallback, T def, String hint, T Function(String) parse}) async {
  assert(T is String || parse != null, '`parse` can only be null if T is String. T is $T');
  stdout.write('$question${hint != null || def != null ? ' (${hint ?? def})' : ''} >>> ');
  // await stdout.flush();
  final ret = await stdinLined.first;
  print('[OUTPUT = $ret]'); // for debugging
  if (ret.isEmpty && fallback != null) return def;
  else if (parse != null) {
    try {
      return parse(ret);
    } on FormatException {
      return fallback;
    }
  }

  return ret as T;
}

Future<void> main() async {
  final name = await prompt<String>('What is your name?');
}
```

Unfortunately I was not able to reproduce the bug.

```
$ dart --version
Dart VM version: 2.6.0 (Thu Oct 24 17:52:22 2019 +0200) on "linux_x64"
$ dart test.dart
What is your name? >>> foo
[OUTPUT = foo]
```

I need more information about your environment to track down and reproduce your bug. You didn't answer these questions in the issue template, can you do that, plus provide whatever else information is useful in reproducing this bug?

```
In your issue, please include:

* Dart SDK Version (`dart --version`)
* Whether you are using Windows, MacOSX, or Linux (if applicable)
* Whether you are using Chrome, Safari, Firefox, Edge (if applicable)

Missing some or all of the above might make the issue take longer or be
impossible to act on.
```

But in any case, thank you for filing a bug and helping improve Dart, I look forward to hearing more details.

**Comment by mpfaff** [comment]
Hi @sortie 

Thanks for testing it. I tried running it again in a terminal and confirmed that it does work. When I went back to debug it in VSCode where I'd originally found the bug, it appeared once again. DartCode 3.7 adds support for debugging Dart programs in a terminal because the debug console doesn't support stdin, and it seems that's where the bug comes from. I guess I'll file a report over there.
